src/oph/ehoks/hoks.clj

```diff
@@ -331,12 +346,17 @@
                       {:type                   :disallowed-update
                        :old-opiskeluoikeus-oid old-opiskeluoikeus-oid
                        :new-opiskeluoikeus-oid new-opiskeluoikeus-oid})))
-    (when (and (some? new-oppija-oid)
-               (not= new-oppija-oid old-oppija-oid))
-      (throw (ex-info "Updating `oppija-oid` in HOKS is not allowed!"
-                      {:type           :disallowed-update
-                       :old-oppija-oid old-oppija-oid
-                       :new-oppija-oid new-oppija-oid})))))
+    (when (oppija-oid-changed? new-oppija-oid old-oppija-oid)
```

Comment:
Kattaakohan tämä nyt kaikki "oikeat" tapaukset? Tein tähän alkuun monimutkaisemman päättelyn, mutta se typistyi sitten tähän 😄

Reply:
Katsotaas, eli jos

* eHOKSissa on uusin oppija-oid (master) ja
  - a. yritetään päivittää slave-oidilla
    - alla koodissa `master-oid` ja `new-oppija-oid` ei mätsää, eli tulee poikkeus eikä päivittämiseen asti edetä.
  - b. päivitys tulee master-oidilla
    - ei tule poikkeusta koska `oppia-oid-changed?` palauttaa `false` oidien ollessa samat, eikä näin ollen myöskään turhaan päivitetä.
* eHOKSissa on vanha oppija-oid (slave) ja
  - c. päivitys tulee samalla slave-oidilla
    - `oppija-oid-changed?` palauttaa `false` oidien ollessa samat, joten päivitykseen asti ei edetä, ts. mitään ei tapahtu.
  - d. päivitys tulee eriävällä slave-oidilla
    - tulee poikkeus, koska on eriävä master-oidista. Itse olin ajatellut että tässä voisi olla oiva tilaisuus päivittää ajantasaisin master-oid eHOKSin tietoihin, mutta en toisaalta osaa sanoa olisiko kyseinen toimenpide OK tehdä tuolla tavalla implisiittisesti :thinking: Voi tietysti OPH:laisilta vielä kysäistä. Tuo sun toteutuksessa tehty toimintatapa on mun mielestä joka tapauksessa hyvä, tosin tuo virheviesti "Updating `oppija-oid` in HOKS is not allowed!" varmaan täytyisi muuttaa kuvaavammaksi, koska oppija-oidin päivitys on tosiaan tietyin ehdoin mahdollista. :smile:
  - e. päivitys tulee master-oidilla
    - mätsää ONR:stä haettuun master-oidiin, ei tule poikkeusta ja uusi oid päivittyy indekseihin.

Eli näyttää mun silmään logiikka OK:lta :+1:  Tuossa yllä listasin uskoakseni kaikki mahdolliset tapaukset.

Reply:
Kiitoksia perusteellisesta tapausten läpikäynnistä! Mäkään en keksi enempää keissejä, joten päättely lienee siis ok. 

Tuohon implisiittiseen päivitykseen suhtaudun vähän kaksijakoisesti. Toisaalta olisi hyvä, että HOKSin data päivittyisi "oikeaan/parempaan" suuntaan, mutta toisaalta jos data muuttuu eriksi kuin mitä käyttäjä on pyytänyt, niin hämmentääkö se käyttäjää? Lähinnä ainakin jos tällaisessa tapauksessa palautetaan 400-virhekoodi, mutta silti data muuttui (ja tallentui) niin sekin voi lisätä hämmennystä. Toisaalta hieman vastaava tilanne voi tapahtua myös ei-transaktionaalisissa päivitystapauksissa, joissa ei voidakaan rollbackata kaikkia API-kutsun sivuvaikutuksia, mutta tällöin usein palautuu 5xx-virhe. Toisaalta nyt tässä oppija-oidin vaihtamistapauksessa kyseessä on käyttäjän tekemä virhe datassa, eli 4xx-koodi, esim [422](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422), voisi olla paikallaan. 🤔

Reply:
Samanlaisia ajatuksia itsellänikin heräsi tuosta, todennäköisesti enemmän hämmennystä aiheuttaisi tuo implisiittinen OID:in päivitys, ja näin jälkikäteen ajateltuna olisi varmaan muutenkin melko kyseenalaista, joten parempi jättää päivittämättä tuossa tapauksessa :smile:  Ja joo, 422-virhekoodi taitaa olla tuohon tilanteeseen sopivin. :+1:

Reply:
Minusta oikea virhekoodi on ihan selkeästi joko 400 tai 409.  Käyttäjä on lähettänyt eri OIDin kuin mikä HOKSissa on sallittua olla.